src/write/buf_write.rs

```diff
@@ -0,0 +1,29 @@
+use std::{
+    io,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+pub(crate) trait AsyncBufWrite {
+    /// Attempt to return an internal buffer to write to, flushing data out to the inner reader if
+    /// it is full.
+    ///
+    /// On success, returns `Poll::Ready(Ok(buf))`.
+    ///
+    /// If the buffer is full and cannot be flushed, the method returns `Poll::Pending` and
```

Comment:
```suggestion
    /// If the buffer's `pending` data cannot be flushed, the method returns `Poll::Pending` and
```

Reply:
This only requires a partial flush, I'll have a think about how to make the semantics a bit clearer.